### PR TITLE
Stopwatch

### PR DESCRIPTION
**Pull request by tdealtry**
Added a new class that calculates min/max/average/total time (and a histogram) for a specific function to use. The idea is that it'd be implemented in most tools, and turned on by the user in the config file. I've implemented it for the template and nhits tools so far, for the Initialise(), Execute(), Finalise() methods

Some questions/comments
- I've added the `Stopwatch` class in DataModel/, but the DataModel doesn't use it. We should have some common area that contains helper classes/functions that all tools can use, but aren't part of the DataModel. As a first suggestion - "Utilities"?
- It uses ROOTs TStopwatch, which is fine, but reports CPU time to an accuracy of 0.01s. It'd be simple enough to swap to another timer - does anyone have a favourite? (note that the "realtime" returns more sig fig)
- This builds on #37

### CODE REVIEW

File: UserTools/template/MyTool.cpp [hunk 1]
```diff
@@ -8,22 +8,78 @@ bool MyTool::Initialise(std::string configfile, DataModel &data){
...
 
   m_data= &data;
 
+
+  /// YOUR CODE HERE
+
+
+  if(m_stopwatch) {
+    m_stopwatch->Stop();
+    ss << "INFO: nhits::Initialise() run stats" << m_stopwatch->Result();
```
Comment (brichards64) [inline UserTools/template/MyTool.cpp:35]: This Should be Mytool not nhits

File: UserTools/template/MyTool.cpp [hunk 2]
```diff
@@ -8,22 +8,78 @@ bool MyTool::Initialise(std::string configfile, DataModel &data){
...
   return true;
 }
 
 
 bool MyTool::Finalise(){
 
+  Log("INFO: MyTool::Finalise() starting", INFO, m_verbose);
+
+  if(m_stopwatch) {
+    ss << "INFO: nhits::Execute() run stats" << m_stopwatch->Result(m_stopwatch_file);
```
Comment (brichards64) [inline UserTools/template/MyTool.cpp:67]: again MY Tool not nhits

File: UserTools/template/MyTool.cpp [hunk 3]
```diff
@@ -8,22 +8,78 @@ bool MyTool::Initialise(std::string configfile, DataModel &data){
...
+    StreamToLog(INFO);
+    m_stopwatch->Reset();
+    m_stopwatch->Start();
+  }
+
+  //// YOUR CODE HERE
+
+  if(m_stopwatch) {
+    m_stopwatch->Stop();
+    ss << "INFO: nhits::Initialise() run stats" << m_stopwatch->Result();
```
Comment (brichards64) [inline UserTools/template/MyTool.cpp:77]: nhits again

**Comment by brichards64** [comment]
Hi Tom, can you fix the issues i commented and resolve the conflicts please then ill merge it

File: DataModel/Stopwatch.cpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+#include <sstream>
+#include <iostream>
+#include <algorithm>
+
+#include "TH1D.h"
+#include "TLegend.h"
+#include "TCanvas.h"
+
+///////////////////////////////////////////////////
+void Stopwatch::Start()
```
Comment (brichards64) [inline DataModel/Stopwatch.cpp:12]: can you make the { inline

File: UserTools/nhits/nhits.cpp [hunk 5]
```diff
@@ -7,20 +7,26 @@ nhits::nhits():Tool(){}
...
+  //Setup and start the stopwatch
+  bool use_stopwatch = false;
+  m_variables.Get("use_stopwatch", use_stopwatch);
+  m_stopwatch = use_stopwatch ? new Stopwatch() : 0;
 
-  m_data->triggeroutput=false;
+  m_stopwatch_file = "";
+  m_variables.Get("stopwatch_file", m_stopwatch_file);
 
+  if(m_stopwatch)
```
Comment (brichards64) [inline UserTools/nhits/nhits.cpp:24]: can these be one line too please

**Comment by tdealtry** [comment]
This is ready to go, as long as you're happy with the name of the new "Utilites" directory

**Comment by brichards64** [comment]
Yeh i'm not sure about utilities....

Whilst i understand it's purpose and its a good idea, it breaks wider compatibility. All of ToolDAq basically has two areas. Tools which are individual to a tool and Datamodel that is common to all tools. So normally it would go in DataModel as is. Whilst Datamodel might be a crap name as it starts to include things that are just common and not just data objects, in all the other deployments of tooldaq utilities have either just been a class in the data model folder or a utilities.h file in the data model that contained lots of useful functions. 

Im not sure we want to diverge it so much as we break compatipility with other deployments and i like the simplicity of one location for all common things.

**Comment by brichards64** [comment]
See here in the latest version of the application template https://github.com/ToolDAQ/ToolDAQApplication/tree/Application/DataModel

that's the route i went down

**Comment by brichards64** [comment]
iv also used algorithms.h or functions.h in datamodel in the past

**Comment by tdealtry** [comment]
Ok, that makes sense since there's 10s of tooldaq repos. I've reverted the Utilities folder commit
I think this should't live in Utilities.h, since it's a separate class (rather than a function). But could rename it UtilStopwatch or similar?

**Comment by brichards64** [comment]
yeh that's a good idea, i like having a prefix that shows its a utility.

name space might work as well, util::stopwatch. Although if we do that we might loose some people who are not familiar with how to use namespaces.

what do you think?

**Comment by tdealtry** [comment]
I've done both, however the class `util::Stopwatch` is in the file `UtilStopwatch.h`, so maybe that's a bit confusing? Just use namespaces do you think?

**Comment by brichards64** [comment]
hmmmmm...... yeh its just if people know how to use namespaces or not. but i think i prefer it as you suggested.

so please chang eit back to Stopwatch.h

**Comment by tdealtry** [comment]
Should be ready to go now

**Comment by tdealtry** [comment]
Back to you to check Ben - I've reduced the code required in the tools. It's not quite as simple as you maybe wanted, but it is a one-liner
(If you know how to call a tools Log method from Stopwatch, we can simplify it more, but that sounds messy... or we could tell stopwatch about DataModel, but that again is a confusing dependency...)

File: UserTools/template/MyTool.cpp [hunk 6]
```diff
@@ -11,19 +11,53 @@ bool MyTool::Initialise(std::string configfile, DataModel &data){
...
+
+  if(m_stopwatch) m_stopwatch->Start();
+
   m_data= &data;
 
+
+  /// YOUR CODE HERE
+
+
+  if(m_stopwatch) Log(m_stopwatch->Result("Initialise"), INFO, verbose);
```
Comment (brichards64) [inline UserTools/template/MyTool.cpp:30]: might be nice to have MyTool:Initialise and the same for finalise and execute. That way if your benchmarking multiple tools you can tell them apart.

**Comment by brichards64** [comment]
other than the above suggestion looks good. There is a way of doing it that wouldn't be too messy using function pointers and alike, but this is more verbose and easy for people to understand so i think its fine as is.

**Comment by brichards64** [comment]
never mind ignore the above i see you ahve the toolname in the constructor. So you had thought of that already

looks good to me then

**Pull request merged**